### ISSUE TITLE
Memory leak with systemd input plugin and DB

### ISSUE THREAD

**Issue by HackToHell** [fluent/fluent-bit#1798]
## Bug Report

When the systemd input plugin is enabled with a DB to store the cursor, there's a memory leak.

**To Reproduce**
```
[SERVICE]
  Flush         1
  Log_Level     info
  Daemon        off
  HTTP_Server   On
  HTTP_Listen   0.0.0.0
  HTTP_Port     2020
  Streams_File  stream_processor.conf
  storage.max_chunks_up 5

[INPUT]
  Name            systemd
  Tag             kube.docker.service
  DB             /tmp/db
  Systemd_Filter  _SYSTEMD_UNIT=docker.service
  alias           journal
  routable        false
```

This was running on a openshift cluster with logs of the kubernetes format. 

I ran valgrind and got the following output.

https://gist.github.com/HackToHell/534ebf66764a0fda36bdd00144582721

**Expected behavior**
With the DB removed, there's no memory leak.

**Your Environment**
* Version used: At commit 407e061c89d3f92014855f235a3aea6c5c8c33cc
* Configuration: Posted above
* Environment name and version: Openshift origin 3.x 
* Server type and version: Docker version 1.13.1, build 07f3374/1.13.1, systemd-219-57.el7
* Operating System and version: Centos 7
* Filters and plugins: None for the test.

**Comment by edsiper** [comment]
I am looking at the code and here _qs.rows_ is only incremented when the callback is invoked:

https://github.com/fluent/fluent-bit/blob/master/plugins/in_systemd/systemd_db.c#L84

after that, the cursor string generated by the callback is released. 

some questions:

- which Fluent Bit version are you running?
- running normally outside of Valgrind, do you see Fluent Bit memory going up without stopping?
- if you remove the stream processor configuration file, do you see the same behavior?

**Comment by HackToHell** [comment]
Normal
![](https://i.imgur.com/oQxBzLc.png)
Immediately after adding the DB cursor. (After dropping the stream processor config)
![](https://i.imgur.com/aT7nLUK.png)

1) Version: At commit https://github.com/fluent/fluent-bit/commit/407e061c89d3f92014855f235a3aea6c5c8c33cc 
2) Yes 
3) Yes

**Comment by edsiper** [comment]
working on this

**Comment by edsiper** [comment]
I was not able to reproduce the issue, anyways I did some improvements on journal cursor management that I would like to see if the issue in your environment goes away.

Would you please re-build Fluent Bit using the __issue-1798__ branch ?, top commit on that branch is ef05f9d5

If you cannot rebuild your docker image let me know and I can provide you one.

**Comment by edsiper** [comment]
@HackToHell any luck with issue-1798 branch ?

**Comment by prakashmishra1598** [comment]
@edsiper , We built and deployed with the **issue-1798** branch. The results were similar with memory usage increasing. 
![Screenshot from 2019-12-19 13-52-09](https://user-images.githubusercontent.com/17315130/71160150-08a48200-226d-11ea-8f92-7f6b2f726a80.png)

**Comment by edsiper** [comment]
hmmm

the volume where you store the database file (DB), is a remote end point like NFS or similar ?

**Comment by HackToHell** [comment]
@edsiper No we tried with two different dbs, one on the container itself, at `/tmp/x` and one more as a hostPath mount.

**Comment by dsbrng25b** [comment]
Do you have duplicates in the in_systemd_cursor table in the DB? I observed a similar problem (#1864 ). In my case I could stop the memory leak by removing the duplicates from the table and restarting fluent-bit.

**Comment by edsiper** [comment]
thanks @HackToHell and @dsbrng25b for providing useful information to address the issue. Despite the root cause of corruption could not be found (non-reproducible), I've pushed a fix that makes sure no leaks will be generated and if a corrupted database is found upon start, it will do the proper sanitization.

Fixed through a1b5344d . The fix will be backported to v1.3 series.
